vcd/resource_vcd_nsxv_firewall_rule_test.go

```diff
@@ -0,0 +1,879 @@
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	action          = "deny"
+	enabled         = "false"
+	logging_enabled = "true"
+
+	source {
+		org_networks = ["${vcd_network_routed.test-routed[0].name}"]
+	}
+  
+	destination {
+		exclude      = "true"
+		org_networks = ["${vcd_network_routed.test-routed[1].name}"]
+	}
+	service {
+		protocol = "tcp"
+		port     = "443"
+	}
+}
+resource "vcd_network_routed" "test-routed" {
+  count        = 2
+  name         = "firewall-test-${count.index}"
+  org          = "{{.Org}}"
+  vdc          = "{{.Vdc}}"
+  edge_gateway = "{{.EdgeGateway}}"
+  gateway      = "10.201.${count.index}.1"
+  netmask      = "255.255.255.0"
+
+  static_ip_pool {
+    start_address = "10.201.${count.index}.10"
+    end_address   = "10.201.${count.index}.20"
+  }
+}
+`
+
+const testAccVcdEdgeFirewallRule4 = `
+resource "vcd_nsxv_firewall_rule" "rule4" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "test-rule-4"
+	action = "deny"
+
+	source {
+		gateway_interfaces = ["internal"]
+	}
+  
+	destination {
+		gateway_interfaces = ["external"]
+	}
+
+	service {
+		protocol = "tcp"
+		port     = "443"
+	}
+	
+	service {
+		protocol = "tcp"
+		port     = "8443"
+		source_port = "20000-40000"
+	}
+
+	service {
+		protocol = "udp"
+		port     = "10000"
+	}
+
+	service {
+		protocol    = "udp"
+		port        = "10000"
+		source_port = "20000"
+	}
+
+	service {
+		protocol = "icmp"
+	}
+  }
+
+data "vcd_nsxv_firewall_rule" "rule4" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+
+	rule_id      = "${vcd_nsxv_firewall_rule.rule4.id}"
+}
+`
+
+const testAccVcdEdgeFirewallRule5 = `
+resource "vcd_nsxv_firewall_rule" "rule5" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "test-rule-5"
+
+	source {
+		gateway_interfaces = ["internal"]
+	}
+  
+	destination {
+		gateway_interfaces = ["external"]
+	}
+
+	service {
+		protocol = "tcp"
+	}
+
+	service {
+		protocol = "udp"
+	}
+
+  }
+`
+
+const testAccVcdEdgeFirewallRule8 = `
+resource "vcd_nsxv_firewall_rule" "rule6" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "below-rule"
+	action = "accept"
+
+	source {
+		ip_addresses = ["10.10.10.0/24", "11.10.10.0/24"]
+	}
+  
+	destination {
+		ip_addresses = ["20.10.10.0/24", "21.10.10.0/24"]
+	}
+
+	service {
+		protocol = "any"
+	}
+}
+
+resource "vcd_nsxv_firewall_rule" "rule6-6" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "above-rule"
+	action = "accept"
+	above_rule_id = "${vcd_nsxv_firewall_rule.rule6.id}"
+
+
+	source {
+		ip_addresses = ["30.10.10.0/24", "31.10.10.0/24"]
+	}
+  
+	destination {
+		ip_addresses = ["40.10.10.0/24", "41.10.10.0/24"]
+	}
+
+	service {
+		protocol = "any"
+	}
+
+	depends_on = ["vcd_nsxv_firewall_rule.rule6"]
```

Comment:
Do we need depends_on if have `above_rule_id = "${vcd_nsxv_firewall_rule.rule6.id}"` ?

Reply:
```suggestion
        # this another way to control rules order if `above_id` isn't used
	depends_on = ["vcd_nsxv_firewall_rule.rule6"]
```

Reply:
Added.